### PR TITLE
[Merged by Bors] - feat: some API for explicit limits in CompHaus

### PR DESCRIPTION
**Pull request by dagurtomas**
Co-authored by: 
- Riccardo Brasca [riccardo.brasca@gmail.com](mailto:riccardo.brasca@gmail.com) @riccardobrasca 
- Filippo A. E. Nuccio [filippo.nuccio@univ-st-etienne.fr](mailto:filippo.nuccio@univ-st-etienne.fr) @faenuccio 

We add some more API for the compatibility of explicit and abstract limits in CompHaus
This work was done during the 2023 Copenhagen masterclass on formalisation of condensed mathematics. Numerous participants contributed.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean [hunk 1]
```diff
@@ -109,6 +109,53 @@ def pullback.isLimit : Limits.IsLimit (pullback.cone f g) :=
...
+      ToExplicit, pullback.lift_snd]
+
+@[simp]
+theorem fromExplicit_comp_toExplicit_eq_id :
+    (FromExplicit f g ≫ ToExplicit f g) = 𝟙 _ :=
+  pullback.hom_ext f g _ _ (by simp [ToExplicit, FromExplicit]) (by simp [ToExplicit, FromExplicit])
+
+@[simps]
+noncomputable
+def FromExplicitIso : CompHaus.pullback f g ≅ Limits.pullback f g where
```
Comment (adamtopaz) [inline Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean:137]: Is all the stuff above really necessary, or can you just define this iso using `pullback.isLimit` and `conePointUniqueUpToIso` and work with that instead? Similarly for finite coproducts.

File: Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean [hunk 2]
```diff
@@ -172,6 +219,94 @@ def finiteCoproduct.isColimit : Limits.IsColimit (finiteCoproduct.cocone X) wher
...
+    (Limits.Sigma.ι X a) ≫ ToFiniteCoproduct X = finiteCoproduct.ι X a := by
+  simp [ToFiniteCoproduct]
+
+@[simp]
+theorem finiteCoproduct.ι_comp_fromFiniteCoproduct (a : α) :
+    (finiteCoproduct.ι X a) ≫ FromFiniteCoproduct X = Limits.Sigma.ι X a := by
+  simp [FromFiniteCoproduct]
+
+@[simps] noncomputable
+def ToFiniteCoproductHomeo : (∐ X : _) ≃ₜ finiteCoproduct X where
```
Comment (adamtopaz) [inline Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean:279]: Don't we have some general way of converting isomorphisms in `CompHaus` to homeomorphisms? Maybe one has to pass through `TopCat`?

File: Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean [hunk 3]
```diff
@@ -172,6 +219,94 @@ def finiteCoproduct.isColimit : Limits.IsColimit (finiteCoproduct.cocone X) wher
...
+
+end Iso
+
+lemma finiteCoproduct.ι_injective (a : α) : Function.Injective (finiteCoproduct.ι X a) := by
+  intro x y hxy
+  exact eq_of_heq (Sigma.ext_iff.mp hxy).2
+
+lemma finiteCoproduct.ι_jointly_surjective (R : finiteCoproduct X) :
+    ∃ (a : α) (r : X a), R = finiteCoproduct.ι X a r := by
+  exact ⟨R.fst, R.snd, rfl⟩
```
Comment (adamtopaz) [inline Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean:302]:
```suggestion
    ∃ (a : α) (r : X a), R = finiteCoproduct.ι X a r :=
  ⟨R.fst, R.snd, rfl⟩
```

File: Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean [hunk 4]
```diff
@@ -172,6 +219,94 @@ def finiteCoproduct.isColimit : Limits.IsColimit (finiteCoproduct.cocone X) wher
...
+
+lemma finiteCoproduct.ι_desc_apply {B : CompHaus} {π : (a : α) → X a ⟶ B} (a : α) :
```
[the remark this answers is not in the document]
Comment (dagurtomas) [reply Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean:294]: Tagging `finiteCoproduct.ι_desc` with `elementwise` gives the following error: 

"elementwise lemma for CompHaus.finiteCoproduct.ι_desc is trivial after applying ConcreteCategory lemmas, which can be caused by how applications are unfolded. Using elementwise is unnecessary." 

Still, `simp` doesn't find anything that proves the analogue of `ι_desc` applied to an element.

File: Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean (hunk 1 above)
Comment (dagurtomas) [reply Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean]: You're right, `conePointUniqueUpToIso worked

File: Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean (hunk 4 above)
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean:217]: Ugh, `elementwise` is clearly buggy. Could you leave a comment along the lines of "`elementwise` should work here, but doesn't", so we can track this down later?

File: Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean (hunk 3 above)
Comment (semorrison) [reply Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean]: @dagurtomas, it's helpful if you click "resolve conversation" when you're confident you've reacted to a review comment sufficiently. It saves later reviewers having to work out if something is still a live issue or not.

File: Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean [hunk 5]
```diff
@@ -36,10 +36,9 @@ pairs `(x,y)` such that `f x = g y`, with the topology induced by the product.
 -/
```
Comment (semorrison) [inline Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean:36]: Ah, I see, this file already existed. I think this probably should just be `Limits.lean`.

File: Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean [hunk 6]
```diff
@@ -109,6 +108,24 @@ def pullback.isLimit : Limits.IsLimit (pullback.cone f g) :=
     (fun _ => pullback.lift_snd _ _ _ _ _)
     (fun _ _ hm => pullback.hom_ext _ _ _ _ (hm .left) (hm .right))
 
+section Isos
+
+noncomputable
+def FromExplicitIso : CompHaus.pullback f g ≅ Limits.pullback f g :=
```
Comment (semorrison) [inline Mathlib/Topology/Category/CompHaus/ExplicitLimits.lean:114]: Shouldn't start uppercase unless it is a Type/Prop.

File: Mathlib/Topology/Category/CompHaus/Limits.lean [hunk 7]
```diff
@@ -109,6 +108,24 @@ def pullback.isLimit : Limits.IsLimit (pullback.cone f g) :=
     (fun _ => pullback.lift_snd _ _ _ _ _)
     (fun _ _ hm => pullback.hom_ext _ _ _ _ (hm .left) (hm .right))
 
+section Isos
+
+noncomputable
+def pullbackIsoPullback : CompHaus.pullback f g ≅ Limits.pullback f g :=
+Limits.IsLimit.conePointUniqueUpToIso (pullback.isLimit f g) (Limits.limit.isLimit _)
+
+theorem explicit_fst_eq :
```
Comment (semorrison) [inline Mathlib/Topology/Category/CompHaus/Limits.lean:117]:
```suggestion
theorem pullback_fst_eq :
```
and below (assuming this is all inside the `CompHaus` namespace).

File: Mathlib/Topology/Category/CompHaus/Limits.lean [hunk 8]
```diff
@@ -172,6 +189,38 @@ def finiteCoproduct.isColimit : Limits.IsColimit (finiteCoproduct.cocone X) wher
...
     exact hm
 
+section Iso
+
+noncomputable
+def coproductIsoCoproduct : finiteCoproduct X ≅ ∐ X :=
+Limits.IsColimit.coconePointUniqueUpToIso (finiteCoproduct.isColimit X) (Limits.colimit.isColimit _)
+
+theorem Sigma.ι_comp_toFiniteCoproduct (a : α) :
+    finiteCoproduct.ι X a = (Limits.Sigma.ι X a) ≫ (coproductIsoCoproduct X).inv := by
```
Comment (semorrison) [inline Mathlib/Topology/Category/CompHaus/Limits.lean:199]: Seems like the sides should be reversed here?

**Comment by semorrison** [comment]
bors d+

**Pull request closed without merging**